src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs

```diff
@@ -112,5 +139,158 @@ private static bool IsFactory<TReturn>(MethodInfo? factory)
             var servicesProperty = hostType.GetProperty("Services", DeclaredOnlyLookup);
             return (IServiceProvider?)servicesProperty?.GetValue(host);
         }
+
+        // This host builder captures calls to the IHostBuilder then replays them on the application's
+        // IHostBuilder when the event fires
+        private class DeferredHostBuilder : IHostBuilder, IObserver<DiagnosticListener>, IObserver<KeyValuePair<string, object?>>
+        {
+            public IDictionary<object, object> Properties { get; } = new Dictionary<object, object>();
+
+            private readonly string[] _args;
+            private readonly Assembly _assembly;
+
+            private readonly TaskCompletionSource<IHost> _hostTcs = new();
+            private IDisposable? _disposable;
+
+            private Action<IHostBuilder> _configure = b => { };
+
+            // The amount of time we wait for the diagnostic source events to fire
+            private static readonly TimeSpan _waitTimeout = TimeSpan.FromSeconds(5);
```

Comment:
I'm not sure what a decent default should be. I had 30 seconds but that would make testing invalid tests kinda awful.